packages/permission-controller/src/PermissionController.ts

```diff
@@ -1993,6 +2036,95 @@ export class PermissionController<
     return approvedRequest as PermissionsRequest;
   }
 
+  /**
+   * Reunites all the side-effects (onPermitted and onFailure) of the requested permissions inside a record of arrays.
+   *
+   * @param permissions - The approved permissions.
+   * @returns The {@link SideEffects} object containing the handlers arrays.
+   */
+  private getSideEffects(permissions: RequestedPermissions) {
+    return Object.keys(permissions).reduce<SideEffects>(
+      (sideEffectList, targetName) => {
+        const targetKey = this.getTargetKey(targetName);
+
+        if (targetKey) {
+          const specification = this.getPermissionSpecification(targetKey);
+
+          if (specification.sideEffect) {
+            sideEffectList.permittedHandlers.push(
+              specification.sideEffect.onPermitted,
+            );
+
+            if (specification.sideEffect.onFailure) {
+              sideEffectList.failureHandlers.push(
+                specification.sideEffect.onFailure,
+              );
+            }
+          }
+        }
+        return sideEffectList;
+      },
+      { permittedHandlers: [], failureHandlers: [] },
+    );
+  }
+
+  /**
+   * Executes the side-effects of the approved permissions while handling the errors if any.
+   * It will pass an instance of the {@link messagingSystem} and the request data associated with the permission request to the handlers through its params.
+   *
+   * @param sideEffects - the side-effect record created by {@link getSideEffects}
+   * @param requestData - the permissions requestData.
+   * @returns the value returned by all the `onPermitted` handlers in an array.
+   */
+  private async executeSideEffects(
+    sideEffects: SideEffects,
+    requestData: PermissionsRequest,
+  ) {
+    const { permittedHandlers, failureHandlers } = sideEffects;
+    const params = {
+      requestData,
+      messagingSystem: this.messagingSystem,
+    };
+
+    const promiseResults = await Promise.allSettled(
+      permittedHandlers.map((permittedHandler) => permittedHandler(params)),
+    );
+
+    // lib.es2020.promise.d.ts does not export its types so we're using a simple type.
+    const rejectedHandlers = promiseResults.filter(
+      (promise) => promise.status === 'rejected',
+    ) as { status: 'rejected'; reason: Error }[];
+
+    if (rejectedHandlers.length > 0) {
+      if (failureHandlers.length > 0) {
+        try {
+          await Promise.all(
+            failureHandlers.map((failureHandler) => failureHandler(params)),
```

Comment:
Currently it seems like we're running all the failure handers if any of the permitted handlers fail, so just one rejection would run all the failure handlers, even for the successful permitted handlers. Is this intentional?

Reply:
Yes this is intentional. Requested permissions can be inter-dependent, same for side-effects. Since permissions are not granted if a side-effect fails `onFailure` will have to rollback any changes made for all the side-effects